Return 404 response from third party auth login when SAML disabled

omarkhan opened the pull request:
See [SOL-1550](https://openedx.atlassian.net/browse/SOL-1550).

The third-party auth handler currently returns a 500 response if the feature is disabled. This pull request modifies it to return a 404 instead.

edx-webhook commented:
Thanks for the pull request, @omarkhan! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=11436&repo=edx%2Fedx-platform

bradenmacdonald commented on common/djangoapps/third_party_auth/saml.py:
```diff
@@ -25,9 +26,6 @@ def setting(self, name, default=None):
             self._config = SAMLConfiguration.current()  # pylint: disable=attribute-defined-outside-init
-        if not self._config.enabled:
-            from django.core.exceptions import ImproperlyConfigured
-            raise ImproperlyConfigured("SAML Authentication is not enabled.")
```
I think this check is a useful safety measure and worth keeping here. The main problem is that by raising a django exception instead of a python-social-auth exception, we cause a 500 instead of a 3xx or 4xx error.

What about
```
        if not self._config.enabled:
            raise SAMLBackendDisabled()
```

where that exception is defined earlier in this file as:
```python
from social.exceptions import SocialAuthBaseException
class SAMLBackendDisabled(SocialAuthBaseException):
    def __str__(self):
        return 'The SAML third party authentication backend is not enabled.'
```

bradenmacdonald commented on common/djangoapps/third_party_auth/saml.py:
```diff
@@ -35,14 +33,18 @@ def setting(self, name, default=None):
         TODO: remove this method once the fix is merged upstream:
         https://github.com/omab/python-social-auth/pull/821
         """
+        if not self._config.enabled:
```
This code differs from the upstream version, and we want to remove this whole method once the upstream patch has been merged. With the change I suggested above, we can remove these two lines, though the result is a 301 redirect instead of a 404 error. Although that's not as good as a 404, I think it may be better in this case since it reduces the difference with upstream. What do you think?

omarkhan commented:
@bradenmacdonald I have taken a different approach, see the last commit. By wrapping the view we get the same result, irrespective of any upstream changes.

bradenmacdonald commented on common/djangoapps/third_party_auth/views.py:
```diff
@@ -61,6 +61,16 @@ def lti_login_and_complete_view(request, backend, *args, **kwargs):
+    auth is disabled.
+    """
+    if not SAMLConfiguration.is_enabled():
+        raise Http404
```
This view is used by other types of third party auth, not just SAML. So this fix would break things like Google logins (`/auth/login/google-oauth2/`) which don't depend on SAMLConfiguration.

omarkhan commented:
@bradenmacdonald good point, I have reverted the change. However I think it makes more sense to keep the change in the `auth_url` method, for 2 reasons:

- It's easier to see what will happen by reading the code (a 404), without having to look into how python social auth handles exceptions
- We ensure that the user gets a 404 response, which makes more sense in this case than a redirect

I have changed to TODO to specify that the specific check should be removed once the fix is merged upstream, rather than the whole method. If you feel strongly about going the custom exception route let me know and I will make the change.

bradenmacdonald commented:
@omarkhan Ok, this is good with me :+1:

omarkhan commented:
@feanil this is ready for review.

You can test that this returns a 404 here: http://pr11436.sandbox.opencraft.com/auth/login/tpa-saml/

feanil commented:
:+1:
